packages/deployments/contracts/contracts/SponsorVault.sol

```diff
@@ -0,0 +1,237 @@
+
+  /**
+   * @notice Sets the Connext.
+   * @dev Connext and sponsor vault store references to each other
+   * @param _connext The address of the Connext implementation
+   */
+  function setConnext(address _connext) external onlyOwner {
+    _setConnext(_connext);
+  }
+
+  /**
+   * @notice Sets default origin domain gas token to this domain gas token rate.
+   * @param _originDomain The origin domain
+   * @param _rate The default rate
+   */
+  function setRate(uint32 _originDomain, Rate calldata _rate) external onlyOwner {
+    if (_originDomain == 0) revert SponsorVault__setRate_invalidOriginDomain();
+
+    rates[_originDomain] = _rate;
+
+    emit RateUpdated(_originDomain, _rate, msg.sender);
+  }
+
+  /**
+   * @notice Sets the maximum sponsored relayer fee amount.
+   * @param _relayerFeeCap The new relayerFeeCap
+   */
+  function setRelayerFeeCap(uint256 _relayerFeeCap) external onlyOwner {
+    relayerFeeCap = _relayerFeeCap;
+
+    emit RelayerFeeCapUpdated(_relayerFeeCap, msg.sender);
+  }
+
+  /**
+   * @notice Sets of an oracle that provides origin domain gas token to this domain gas token rates.
+   * @param _gasTokenOracle The oracle address
+   */
+  function setGasTokenOracle(address _gasTokenOracle) external onlyOwner {
+    gasTokenOracle = _gasTokenOracle;
+
+    emit GasTokenOracleUpdated(_gasTokenOracle, msg.sender);
+  }
+
+  /**
+   * @notice Sets the address of an exchange used for swapping this domain gas token for a given token.
+   * @param _token The address of the token
+   * @param _tokenExchange The oracle of the exchange
+   */
+  function setTokenExchange(address _token, address payable _tokenExchange) external onlyOwner {
+    if (_token == address(0)) revert SponsorVault__setTokenExchange_invalidAdopted();
+    tokenExchanges[_token] = _tokenExchange;
+
+    emit TokenExchangeUpdated(_token, _tokenExchange, msg.sender);
+  }
+
+  // ============ External functions ============
+
+  receive() external payable {}
+
+  /**
+   * @notice Performs liquidity fee reimbursement.
+   * @dev Uses the token exchange or liquidity deposited in this contract.
+   * @param _token The address of the token
+   * @param _liquidityFee The liquidity fee amount
+   * @return Sponsored liquidity fee amount
+   */
+  function reimburseLiquidityFees(address _token, uint256 _liquidityFee) external override onlyConnext returns (uint256) {
+    // do not sponsor liquidity fee when there is no token exchange nor liquidity for the diven token
+    if (tokenExchanges[_token] == address(0) && IERC20(_token).balanceOf(address(this)) < _liquidityFee) return 0;
+
+    if (tokenExchanges[_token] != address(0)) {
+      ITokenExchange tokenExchange = ITokenExchange(tokenExchanges[_token]);
+      uint256 amountIn = tokenExchange.getInGivenExactOut(_token, _liquidityFee);
+
+      if (address(this).balance < amountIn) return 0;
+
+      tokenExchange.swapExactOut{value: amountIn}(_token, _liquidityFee, msg.sender);
+    } else {
```

Comment:
`ITokenExchange` would be the external dexs like Uniswap? If so, we have to use their interface. And i think  `swapExactOut` function has to have `amountOutMin` due to slippage.

Reply:
see https://github.com/connext/nxtp/pull/1069#discussion_r865047730

Reply:
and we have to build a contract for token exchange like a kind of aggregator? or we will use existing dexs like uniswap?

Reply:
> and we have to build a contract for token exchange like a kind of aggregator? or we will use existing dexs like uniswap?

I would say the `tokenExchange` would be an `adaptor` for existing DEXs, if not we would need to choose some specific DEX and use its interface, right?

Reply:
I think using the adapter pattern makes more sense -- there are different dexes with different liquidity depths depending on the swap you're doing so we shouldn't force a specific implementation